cellarium/ml/core/module.py

```diff
@@ -114,17 +153,50 @@ def configure_model(self) -> None:
                 )
             )
         else:
-            transforms = None
+            transforms = tuple()
 
-        self.pipeline = CellariumPipeline(transforms)
         if model is None:
             raise ValueError(f"`model` must be an instance of {CellariumModel}. Got {model}")
-        self.pipeline.append(model)
+        self.pipeline = CellariumPipeline(cpu_transforms + transforms + (model,))  # the full pipeline
+        self.module_pipeline = self.pipeline  # training loop pipeline: initially the full pipeline
 
         if not self.hparams["is_initialized"]:
             model.reset_parameters()
             self.hparams["is_initialized"] = True
 
+        # move the cpu_transforms to the dataloader's collate_fn if the dataloader is going to apply them
+        try:
+            if hasattr(self, "trainer"):  # for some reason this check raises a RuntimeError rather than returning False
+                if hasattr(self.trainer, "datamodule"):
```

Comment:
Seems it is safe to use `trainer.datamodule` per our conversation.

Reply:
Here I think I may have to keep the `hasattr(self.trainer, "datamodule")` check for mypy to pass.  Without it, for example if I try
```python
        # move the cpu_transforms to the dataloader's collate_fn if the dataloader is going to apply them
        if self._trainer is not None:
            # if hasattr(self.trainer, "datamodule"):
            if isinstance(self.trainer.datamodule, CellariumAnnDataDataModule):
                self._lightning_training_using_datamodule = True
                self.trainer.datamodule.collate_fn = FunctionComposer(
                    first_applied=collate_fn,
                    second_applied=self.cpu_transforms,
                )
                self.remove_cpu_transforms_from_module_pipeline()
```

I get this from mypy:
```pytb
error: "Trainer" has no attribute "datamodule"  [attr-defined]
```

Reply:
I think you can use either hasattr:

```py
if hasattr(self.trainer, "datamodule") and isinstance(self.trainer.datamodule, CellariumAnnDataDataModule):
```

or type ignore + walrus operator (to avoid additional type ignore)

```py
if isinstance(datamodule := self.trainer.datamodule, CellariumAnnDataDataModule):   # type: ignore[attr-defined]
    self._lightning_training_using_datamodule = True
    datamodule.collate_fn = FunctionComposer(
          first_applied=collate_fn,
          second_applied=self.cpu_transforms,
    )
```

Either should work.